src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs

```diff
@@ -0,0 +1,295 @@
+// </copyright>
+//---------------------------------------------------------------------
+
+#if NETCOREAPP
+using System;
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.OData.Json
+{
+    /// <summary>
+    /// Represents a Utf8JsonWriter for writing JSON data to a stream.
+    /// </summary>
+    internal sealed partial class ODataUtf8JsonWriter : IJsonWriter, IDisposable, IAsyncDisposable
+    {
+        /// <summary>
+        /// Starts a scope for writing a stream value.
+        /// </summary>
+        /// <returns>A stream for writing the stream value.</returns>
+        public Stream StartStreamValueScope()
+        {
+            this.CommitUtf8JsonWriterContentsToBuffer();
+
+            this.WriteItemWithSeparatorIfNeeded();
+
+            this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+            this.Flush();
+
+            this.binaryValueStream =  new ODataUtf8JsonWriterStream(this);
+
+            return this.binaryValueStream;
+        }
+
+        /// <summary>
+        /// Ends a scope for writing a stream value.
+        /// </summary>
+        public void EndStreamValueScope()
+        {
+            this.binaryValueStream.Flush();
+            this.binaryValueStream.Dispose();
+            this.binaryValueStream = null;
+            this.Flush();
+
+            this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+        }
+
+        /// <summary>
+        /// Asynchronously starts a scope for writing a stream value.
+        /// </summary>
+        /// <returns>A task representing the asynchronous operation. The task result contains a stream for writing the stream value.</returns>
+        public async Task<Stream> StartStreamValueScopeAsync()
+        {
+            this.WriteSeparatorIfNecessary();
+            this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+            await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+
+            this.binaryValueStream = new ODataUtf8JsonWriterStream(this);
+
+            return this.binaryValueStream;
+        }
+
+        /// <summary>
+        /// Asynchronously ends a scope for writing a stream value.
+        /// </summary>
+        /// <returns>A task representing the asynchronous operation.</returns>
+        public async Task EndStreamValueScopeAsync()
+        {
+            await this.binaryValueStream.FlushAsync();
+            await this.binaryValueStream.DisposeAsync();
+            this.binaryValueStream = null;
+            await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+
+            this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+        }
+
+        /// <summary>
+        /// Represents a stream for writing UTF-8 JSON data to an ODataUtf8JsonWriter.
+        /// </summary>
+        internal class ODataUtf8JsonWriterStream : Stream
+        {
+            ODataUtf8JsonWriter jsonWriter = null;

```

Comment:
```suggestion
            private readonly ODataUtf8JsonWriter jsonWriter = null;
```
To guarantee the class member can only be initialized from the constructor.